### ISSUE TITLE
node-sass 3.0.0-beta on freebsd 10.0

### ISSUE THREAD

**Issue by vladyslav2** [sass/node-sass#892]
Hello.
I'm getting error while trying to compile node sass for FreeBSD,
Does anybody know how to fix that ? 
thanks

```
root@computer#: env CC=clang CXX=clang++ LINK=clang++ SKIP_SASS_BINARY_DOWNLOAD_FOR_CI=1 V=1 npm -g install node-sass#v3.0.0-alpha.1
/usr/local/bin/node-sass -> /usr/local/lib/node_modules/node-sass/bin/node-sass

> node-sass@2.1.1 install /usr/local/lib/node_modules/node-sass
> node scripts/install.js

Skipping downloading binaries on CI builds

> node-sass@2.1.1 postinstall /usr/local/lib/node_modules/node-sass
> node scripts/build.js

gyp WARN EACCES user "root" does not have permission to access the dev dir "/root/.node-gyp/0.10.30"
gyp WARN EACCES attempting to reinstall using temporary dev dir "/usr/local/lib/node_modules/node-sass/.node-gyp"
gmake: вход в каталог «/usr/local/lib/node_modules/node-sass/build»
  clang++ '-D_LARGEFILE_SOURCE' '-D_FILE_OFFSET_BITS=64' '-DBUILDING_NODE_EXTENSION' -I/usr/local/lib/node_modules/node-sass/.node-gyp/0.10.30/src -I/usr/local/lib/node_modules/node-sass/.node-gyp/0.10.30/deps/uv/include -I/usr/local/lib/node_modules/node-sass/.node-gyp/0.10.30/deps/v8/include -I../node_modules/nan  -fPIC -Wall -Wextra -Wno-unused-parameter -pthread -m64 -O2 -fno-strict-aliasing -fno-omit-frame-pointer -fno-rtti -std=c++0x -fexceptions -frtti -MMD -MF ./Release/.deps/Release/obj.target/binding/src/binding.o.d.raw  -c -o Release/obj.target/binding/src/binding.o ../src/binding.cpp
In file included from ../src/binding.cpp:1:
../node_modules/nan/nan.h:143:5: error: use of undeclared identifier 'abort'
    abort();
    ^
../node_modules/nan/nan.h:152:5: error: use of undeclared identifier 'abort'
    abort();
    ^
2 errors generated.
binding.target.mk:125: ошибка выполнения рецепта для цели «Release/obj.target/binding/src/binding.o»
gmake: *** [Release/obj.target/binding/src/binding.o] Ошибка 1
gmake: выход из каталога «/usr/local/lib/node_modules/node-sass/build»
gyp ERR! build error 
gyp ERR! stack Error: `gmake` failed with exit code: 2
gyp ERR! stack     at ChildProcess.onExit (/usr/local/lib/node_modules/node-sass/node_modules/pangyp/lib/build.js:271:23)
gyp ERR! stack     at ChildProcess.emit (events.js:98:17)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (child_process.js:810:12)
gyp ERR! System FreeBSD 10.0-RELEASE-p9
gyp ERR! command "/usr/local/bin/node" "/usr/local/lib/node_modules/node-sass/node_modules/pangyp/bin/node-gyp" "rebuild"
gyp ERR! cwd /usr/local/lib/node_modules/node-sass
gyp ERR! node -v v0.10.30
gyp ERR! pangyp -v v2.1.0
gyp ERR! not ok 
Build failed
node-sass@2.1.1 /usr/local/lib/node_modules/node-sass
├── get-stdin@4.0.1
├── object-assign@2.0.0
├── replace-ext@0.0.1
├── nan@1.8.2
├── semver@4.3.3
├── shelljs@0.3.0
├── cross-spawn@0.2.9 (lru-cache@2.6.1)
├── mkdirp@0.5.0 (minimist@0.0.8)
├── chalk@0.5.1 (ansi-styles@1.1.0, escape-string-regexp@1.0.3, supports-color@0.2.0, strip-ansi@0.3.0, has-ansi@0.1.0)
├── meow@3.1.0 (camelcase-keys@1.0.0, minimist@1.1.1, indent-string@1.2.1)
├── sass-graph@1.3.0 (commander@2.8.0, lodash@2.4.1, glob@4.5.3)
├── gaze@0.5.1 (globule@0.1.0)
├── npmconf@2.1.1 (uid-number@0.0.5, inherits@2.0.1, osenv@0.1.0, ini@1.3.3, once@1.3.1, nopt@3.0.1, config-chain@1.1.8)
├── mocha@2.2.4 (escape-string-regexp@1.0.2, supports-color@1.2.1, diff@1.0.8, growl@1.8.1, commander@2.3.0, glob@3.2.3, jade@0.26.3, debug@2.0.0)
├── request@2.55.0 (caseless@0.9.0, json-stringify-safe@5.0.0, aws-sign2@0.5.0, forever-agent@0.6.1, stringstream@0.0.4, oauth-sign@0.6.0, tunnel-agent@0.4.0, isstream@0.1.2, node-uuid@1.4.3, qs@2.4.1, combined-stream@0.0.7, form-data@0.2.0, mime-types@2.0.10, http-signature@0.10.1, bl@0.9.4, tough-cookie@0.13.0, har-validator@1.6.1, hawk@2.3.1)
└── pangyp@2.1.0 (which@1.0.9, osenv@0.1.0, rimraf@2.2.8, graceful-fs@3.0.6, nopt@3.0.1, semver@4.2.2, fstream@1.0.4, glob@4.3.5, minimatch@2.0.4, tar@1.0.3, request@2.51.0, npmlog@1.0.0)

```

**Comment by xzyfer** [comment]
Hey @vladyslav2 node-sass@pre is out of date. Please try again with node-sass@beta.

**Comment by vladyslav2** [comment]
Same result,
 maybe i'm doing something wrong ?

And i was wrong - i'm using FreeBSD 10.0


```
root@computer# env CC=clang CXX=clang++ LINK=clang++ SKIP_SASS_BINARY_DOWNLOAD_FOR_CI=1 V=1 npm -g install node-sass#v3.0.0-beta
/usr/local/bin/node-sass -> /usr/local/lib/node_modules/node-sass/bin/node-sass

> node-sass@2.1.1 install /usr/local/lib/node_modules/node-sass
> node scripts/install.js

Skipping downloading binaries on CI builds

> node-sass@2.1.1 postinstall /usr/local/lib/node_modules/node-sass
> node scripts/build.js

gyp WARN EACCES user "root" does not have permission to access the dev dir "/root/.node-gyp/0.10.30"
gyp WARN EACCES attempting to reinstall using temporary dev dir "/usr/local/lib/node_modules/node-sass/.node-gyp"
gmake: Entering directory '/usr/local/lib/node_modules/node-sass/build'
  clang++ '-D_LARGEFILE_SOURCE' '-D_FILE_OFFSET_BITS=64' '-DBUILDING_NODE_EXTENSION' -I/usr/local/lib/node_modules/node-sass/.node-gyp/0.10.30/src -I/usr/local/lib/node_modules/node-sass/.node-gyp/0.10.30/deps/uv/include -I/usr/local/lib/node_modules/node-sass/.node-gyp/0.10.30/deps/v8/include -I../node_modules/nan  -fPIC -Wall -Wextra -Wno-unused-parameter -pthread -m64 -O2 -fno-strict-aliasing -fno-omit-frame-pointer -fno-rtti -std=c++0x -fexceptions -frtti -MMD -MF ./Release/.deps/Release/obj.target/binding/src/binding.o.d.raw  -c -o Release/obj.target/binding/src/binding.o ../src/binding.cpp
In file included from ../src/binding.cpp:1:
../node_modules/nan/nan.h:143:5: error: use of undeclared identifier 'abort'
    abort();
    ^
../node_modules/nan/nan.h:152:5: error: use of undeclared identifier 'abort'
    abort();
    ^
2 errors generated.
binding.target.mk:125: recipe for target 'Release/obj.target/binding/src/binding.o' failed
gmake: *** [Release/obj.target/binding/src/binding.o] Error 1
gmake: Leaving directory '/usr/local/lib/node_modules/node-sass/build'
gyp ERR! build error 
gyp ERR! stack Error: `gmake` failed with exit code: 2
gyp ERR! stack     at ChildProcess.onExit (/usr/local/lib/node_modules/node-sass/node_modules/pangyp/lib/build.js:271:23)
gyp ERR! stack     at ChildProcess.emit (events.js:98:17)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (child_process.js:810:12)
gyp ERR! System FreeBSD 10.0-RELEASE-p9
gyp ERR! command "/usr/local/bin/node" "/usr/local/lib/node_modules/node-sass/node_modules/pangyp/bin/node-gyp" "rebuild"
gyp ERR! cwd /usr/local/lib/node_modules/node-sass
gyp ERR! node -v v0.10.30
gyp ERR! pangyp -v v2.1.0
gyp ERR! not ok 
Build failed
node-sass@2.1.1 /usr/local/lib/node_modules/node-sass
├── get-stdin@4.0.1
├── object-assign@2.0.0
├── replace-ext@0.0.1
├── nan@1.8.2
├── semver@4.3.3
├── shelljs@0.3.0
├── cross-spawn@0.2.9 (lru-cache@2.6.1)
├── mkdirp@0.5.0 (minimist@0.0.8)
├── chalk@0.5.1 (ansi-styles@1.1.0, escape-string-regexp@1.0.3, supports-color@0.2.0, has-ansi@0.1.0, strip-ansi@0.3.0)
├── meow@3.1.0 (minimist@1.1.1, camelcase-keys@1.0.0, indent-string@1.2.1)
├── npmconf@2.1.1 (uid-number@0.0.5, inherits@2.0.1, osenv@0.1.0, ini@1.3.3, once@1.3.1, config-chain@1.1.8, nopt@3.0.1)
├── gaze@0.5.1 (globule@0.1.0)
├── sass-graph@1.3.0 (commander@2.8.0, glob@4.5.3, lodash@2.4.1)
├── mocha@2.2.4 (escape-string-regexp@1.0.2, supports-color@1.2.1, diff@1.0.8, growl@1.8.1, commander@2.3.0, debug@2.0.0, glob@3.2.3, jade@0.26.3)
├── request@2.55.0 (caseless@0.9.0, json-stringify-safe@5.0.0, aws-sign2@0.5.0, forever-agent@0.6.1, stringstream@0.0.4, oauth-sign@0.6.0, tunnel-agent@0.4.0, isstream@0.1.2, node-uuid@1.4.3, qs@2.4.1, combined-stream@0.0.7, form-data@0.2.0, mime-types@2.0.10, http-signature@0.10.1, bl@0.9.4, tough-cookie@0.13.0, hawk@2.3.1, har-validator@1.6.1)
└── pangyp@2.1.0 (which@1.0.9, osenv@0.1.0, rimraf@2.2.8, graceful-fs@3.0.6, nopt@3.0.1, semver@4.2.2, fstream@1.0.4, minimatch@2.0.4, glob@4.3.5, tar@1.0.3, request@2.51.0, npmlog@1.0.0)
```

**Comment by saper** [comment]
While nan should really include <stdlib.h> there is something strange in your setup. 

Can you try installing without -g as a non-root user? If that works: add NPM_CONFIG_UNSAFE_PERM=1 to your env line.

can you remove all /usr/local/lib/node_modules/node-sass? Can you also remove your NPM cache?

**Comment by saper** [comment]
I have a FreeBSD port at http://github.com/saper/ports-exp if you feel brave to experiment.

**Comment by xzyfer** [comment]
It's looks you've tried to install 2.1.1 not beta.7. Please try with this command.

```sh
npm install node-sass@beta
```

**Comment by xzyfer** [comment]
Also can I ask why you're skipping the binary download? We provide node-sass binaries for FreeBSD.

**Comment by vladyslav2** [comment]
@saper same result as non-root user or as root user, with or without -g option
removed /usr/local/lib/node_modules/node-sass - that did not help
how can i clear NPM cache ?

**Comment by vladyslav2** [comment]
@xzyfer: 

```
root@computer:~/temp1# npm -v              
1.4.15
root@computer:~/temp1# node -v
v0.10.30
root@computer:~/temp1# node -p process.versions
{ http_parser: '1.0',
  node: '0.10.30',
  v8: '3.14.5.9',
  ares: '1.9.0-DEV',
  uv: '0.10.28',
  zlib: '1.2.8',
  modules: '11',
  openssl: '1.0.1h' }
root@computer:~/temp1# node -p process.platform
freebsd
root@computer:~/temp1# node -p process.arch
x64
root@computer:~/temp1# npm install node-sass

npm WARN package.json node@0.0.0 No description
npm WARN package.json node@0.0.0 No repository field.
/
> node-sass@2.1.1 install /root/node_modules/node-sass
> node scripts/install.js


module.js:340
    throw err;
          ^
Error: Cannot find module '/root/node_modules/node-sass/scripts/install.js'
    at Function.Module._resolveFilename (module.js:338:15)
    at Function.Module._load (module.js:280:25)
    at Function.Module.runMain (module.js:497:10)
    at startup (node.js:119:16)
    at node.js:906:3
npm ERR! node-sass@2.1.1 install: `node scripts/install.js`
npm ERR! Exit status 8
npm ERR! 
npm ERR! Failed at the node-sass@2.1.1 install script.
npm ERR! This is most likely a problem with the node-sass package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     node scripts/install.js
npm ERR! You can get their info via:
npm ERR!     npm owner ls node-sass
npm ERR! There is likely additional logging output above.

npm ERR! System FreeBSD 10.0-RELEASE-p9
npm ERR! command "node" "/usr/local/bin/npm" "install" "node-sass"
npm ERR! cwd /root/temp1
npm ERR! node -v v0.10.30
npm ERR! npm -v 1.4.15
npm ERR! code ELIFECYCLE
npm ERR! 
npm ERR! Additional logging details can be found in:
npm ERR!     /root/temp1/npm-debug.log
npm ERR! not ok code 0
```

what i'm doing wrong ?

**Comment by saper** [comment]
I noticed you are using ancient npm. Can you install npm from FreeBSD ports?

Your cache as root might be in the `/root/.npm` directory

Don't try `npm install /some/other/directory` until we solve your compilation problem, it won't work.

If you say the result as non-root is "the same", are you still getting those messages:

````
gyp WARN EACCES user "root" does not have permission to access the dev dir "/root/.node-gyp/0.10.30"
gyp WARN EACCES attempting to reinstall using temporary dev dir "/usr/local/lib/node_modules/node-sass/.node-gyp"
````


Can you just:

1. upgrade npm
2. remove npm cache `rm -rf /root/.npm`
3. remove node gyp cache `rm -rf /root/.node-gyp`
4. remove node-sass global installation `rm -rf /usr/local/lib/node_modules/node-sass`
5. try as non-root user:
6. remove npm cache `rm -rf $HOME/.npm`
7. remove node gyp cache `rm -rf $HOME/.node-gyp`
8. try this: `npm install node-sass@beta`


@xzyfer With 2.1.1 we don't have binaries for FreeBSD.

**Comment by vladyslav2** [comment]
@saper:

1-4:
```
root@computer:~/temp1# npm -v     
2.0.2
root@computer:~/temp1# node -v
v0.10.33
root@computer:~/temp1# node -p process.versions
{ http_parser: '1.0',
  node: '0.10.33',
  v8: '3.14.5.9',
  ares: '1.9.0-DEV',
  uv: '0.10.29',
  zlib: '1.2.8',
  modules: '11',
  openssl: '1.0.1j' }

root@computer:~/temp1# npm install node-sass@beta
npm WARN package.json node@0.0.0 No description
npm WARN package.json node@0.0.0 No repository field.
\
> node-sass@3.0.0-beta.7 install /root/node_modules/node-sass
> node scripts/install.js


module.js:340
    throw err;
          ^
Error: Cannot find module '/root/node_modules/node-sass/scripts/install.js'
    at Function.Module._resolveFilename (module.js:338:15)
    at Function.Module._load (module.js:280:25)
    at Function.Module.runMain (module.js:497:10)
    at startup (node.js:119:16)
    at node.js:906:3

npm ERR! FreeBSD 10.0-RELEASE-p9
npm ERR! argv "node" "/usr/local/bin/npm" "install" "node-sass@beta"
npm ERR! node v0.10.33
npm ERR! npm  v2.0.2
npm ERR! code ELIFECYCLE
npm ERR! node-sass@3.0.0-beta.7 install: `node scripts/install.js`
npm ERR! Exit status 8
npm ERR! 
npm ERR! Failed at the node-sass@3.0.0-beta.7 install script.
npm ERR! This is most likely a problem with the node-sass package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     node scripts/install.js
npm ERR! You can get their info via:
npm ERR!     npm owner ls node-sass
npm ERR! There is likely additional logging output above.

```

**Comment by vladyslav2** [comment]
Compiling, right now give a lot of output, and compiling for 3-4 minutes, then:

```
root@computer:~/temp1# env CC=clang CXX=clang++ LINK=clang++ SKIP_SASS_BINARY_DOWNLOAD_FOR_CI=1 V=1 npm -g install node-sass@beta
.....
....
....

rm -rf "Release/sass.a" && cp -af "Release/obj.target/src/sass.a" "Release/sass.a"
  clang++ '-D_LARGEFILE_SOURCE' '-D_FILE_OFFSET_BITS=64' '-DBUILDING_NODE_EXTENSION' -I/usr/local/lib/node_modules/node-sass/.node-gyp/0.10.33/src -I/usr/local/lib/node_modules/node-sass/.node-gyp/0.10.33/deps/uv/include -I/usr/local/lib/node_modules/node-sass/.node-gyp/0.10.33/deps/v8/include -I../node_modules/nan -I../src/libsass  -fPIC -Wall -Wextra -Wno-unused-parameter -pthread -m64 -O2 -fno-strict-aliasing -fno-omit-frame-pointer -fno-rtti -fno-exceptions -fexceptions -std=c++0x -MMD -MF ./Release/.deps/Release/obj.target/binding/src/binding.o.d.raw  -c -o Release/obj.target/binding/src/binding.o ../src/binding.cpp
In file included from ../src/binding.cpp:1:
../node_modules/nan/nan.h:143:5: error: use of undeclared identifier 'abort'
    abort();
    ^
../node_modules/nan/nan.h:152:5: error: use of undeclared identifier 'abort'
    abort();
    ^
../src/binding.cpp:13:25: warning: unused variable 'ctx_w' [-Wunused-variable]
  sass_context_wrapper* ctx_w = static_cast<sass_context_wrapper*>(cookie);
                        ^
1 warning and 2 errors generated.
binding.target.mk:105: recipe for target 'Release/obj.target/binding/src/binding.o' failed
gmake: *** [Release/obj.target/binding/src/binding.o] Error 1
gmake: Leaving directory '/usr/local/lib/node_modules/node-sass/build'
gyp ERR! build error 
gyp ERR! stack Error: `gmake` failed with exit code: 2
gyp ERR! stack     at ChildProcess.onExit (/usr/local/lib/node_modules/node-sass/node_modules/pangyp/lib/build.js:271:23)
gyp ERR! stack     at ChildProcess.emit (events.js:98:17)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (child_process.js:810:12)
gyp ERR! System FreeBSD 10.0-RELEASE-p9
gyp ERR! command "/usr/local/bin/node" "/usr/local/lib/node_modules/node-sass/node_modules/pangyp/bin/node-gyp" "rebuild" "--libsass_ext=" "--libsass_cflags=" "--libsass_ldflags=" "--libsass_library="
gyp ERR! cwd /usr/local/lib/node_modules/node-sass
gyp ERR! node -v v0.10.33
gyp ERR! pangyp -v v2.1.0
gyp ERR! not ok 
Build failed

npm ERR! FreeBSD 10.0-RELEASE-p9
npm ERR! argv "node" "/usr/local/bin/npm" "-g" "install" "node-sass@beta"
npm ERR! node v0.10.33
npm ERR! npm  v2.0.2
npm ERR! code ELIFECYCLE
npm ERR! node-sass@3.0.0-beta.7 postinstall: `node scripts/build.js`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the node-sass@3.0.0-beta.7 postinstall script.
npm ERR! This is most likely a problem with the node-sass package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     node scripts/build.js
npm ERR! You can get their info via:
npm ERR!     npm owner ls node-sass
npm ERR! There is likely additional logging output above.
```

**Comment by vladyslav2** [comment]
As regular user:

```
vladka@computer:~/temp$ npm install node-sass@beta
\
> node-sass@3.0.0-beta.7 install /home/vladka/temp/node_modules/node-sass
> node scripts/install.js

Binary downloaded and installed at /home/vladka/temp/node_modules/node-sass/vendor/freebsd-x64-11/binding.node

> node-sass@3.0.0-beta.7 postinstall /home/vladka/temp/node_modules/node-sass
> node scripts/build.js

` /home/vladka/temp/node_modules/node-sass/vendor/freebsd-x64-11/binding.node ` exists. 
 testing binary.
Binary is fine; exiting.
node-sass@3.0.0-beta.7 node_modules/node-sass
├── get-stdin@4.0.1
├── async-foreach@0.1.3
├── nan@1.8.2
├── mkdirp@0.5.0 (minimist@0.0.8)
├── chalk@1.0.0 (escape-string-regexp@1.0.3, ansi-styles@2.0.1, supports-color@1.3.1, strip-ansi@2.0.1, has-ansi@1.0.3)
├── meow@3.1.0 (object-assign@2.0.0, minimist@1.1.1, camelcase-keys@1.0.0, indent-string@1.2.1)
├── gaze@0.5.1 (globule@0.1.0)
├── glob@5.0.5 (path-is-absolute@1.0.0, inherits@2.0.1, once@1.3.1, inflight@1.0.4, minimatch@2.0.4)
├── npmconf@2.1.1 (uid-number@0.0.5, inherits@2.0.1, osenv@0.1.0, ini@1.3.3, once@1.3.1, config-chain@1.1.8, nopt@3.0.1, semver@4.3.3)
├── sass-graph@1.3.0 (lodash@2.4.1, commander@2.8.1, glob@4.5.3)
├── request@2.55.0 (caseless@0.9.0, json-stringify-safe@5.0.0, aws-sign2@0.5.0, forever-agent@0.6.1, stringstream@0.0.4, oauth-sign@0.6.0, tunnel-agent@0.4.0, isstream@0.1.2, node-uuid@1.4.3, qs@2.4.1, combined-stream@0.0.7, form-data@0.2.0, mime-types@2.0.10, http-signature@0.10.1, bl@0.9.4, tough-cookie@0.13.0, hawk@2.3.1, har-validator@1.6.1)
└── pangyp@2.1.0 (which@1.0.9, osenv@0.1.0, rimraf@2.2.8, graceful-fs@3.0.6, nopt@3.0.1, fstream@1.0.4, semver@4.2.2, glob@4.3.5, minimatch@2.0.4, tar@1.0.3, npmlog@1.0.0, request@2.51.0)
```

```
vladka@computer:~/temp$ npm list | grep sass
└─┬ node-sass@3.0.0-beta.7
  └─┬ sass-graph@1.3.0

```

### Looks like binary node-sass@beta for regular user was installed correctly after updating npm, but installing from root still giving error:

```
root@computer:~/temp1# rm -rf ~/.npm
root@computer:~/temp1# rm -rf ~/.node-gyp
root@computer:~/temp1# rm -rf /root/temp1/node_modules/node-sass
root@computer:~/temp1# rm -rf /usr/local/lib/node_modules/node-sass
root@computer:~/temp1# npm install node-sass@beta
|
> node-sass@3.0.0-beta.7 install /root/temp1/node_modules/node-sass
> node scripts/install.js


module.js:340
    throw err;
          ^
Error: Cannot find module '/root/temp1/node_modules/node-sass/scripts/install.js'
    at Function.Module._resolveFilename (module.js:338:15)
    at Function.Module._load (module.js:280:25)
    at Function.Module.runMain (module.js:497:10)
    at startup (node.js:119:16)
    at node.js:906:3

module.js:340
    throw err;
          ^
```

**Comment by vladyslav2** [comment]
PS try to run grunt from regular user:

```
vladka@computer:~/files/markup/exm$ grunt 
Loading "sass.js" tasks...ERROR
>> Error: `libsass` bindings not found. Try reinstalling `node-sass`?
Warning: Task "sass" not found. Use --force to continue.

Aborted due to warnings.
```

**Comment by saper** [comment]
We are not using grunt for anything, only pangyp to build C++ binary.

You are not supposed to run npm as root (I know, there is no way to do -g otherwise). You might try adding NPM_CONFIG_UNSAFE_PERM=1 in your environment.

**Comment by saper** [comment]
@vladyslav2 I am trying with node v0.12.0 and npm 2.8.1:

````
root@m:~/temp # rm -rf /root/node_modules
root@m:~/temp # npm install node-sass@beta
> node-sass@3.0.0-beta.7 install /root/temp/node_modules/node-sass
> node scripts/install.js

Binary downloaded and installed at /root/temp/node_modules/node-sass/vendor/freebsd-x64-14/binding.node

> node-sass@3.0.0-beta.7 postinstall /root/temp/node_modules/node-sass
> node scripts/build.js

` /root/temp/node_modules/node-sass/vendor/freebsd-x64-14/binding.node ` exists. 
 testing binary.
Binary is fine; exiting.
node-sass@3.0.0-beta.7 node_modules/node-sass
├── get-stdin@4.0.1
├── async-foreach@0.1.3
├── nan@1.8.2
├── chalk@1.0.0 (escape-string-regexp@1.0.3, ansi-styles@2.0.1, supports-color@1.3.1, strip-ansi@2.0.1, has-ansi@1.0.3)
├── mkdirp@0.5.0 (minimist@0.0.8)
├── meow@3.1.0 (object-assign@2.0.0, camelcase-keys@1.0.0, minimist@1.1.1, indent-string@1.2.1)
├── glob@5.0.5 (path-is-absolute@1.0.0, inherits@2.0.1, once@1.3.1, inflight@1.0.4, minimatch@2.0.4)
├── npmconf@2.1.1 (uid-number@0.0.5, inherits@2.0.1, osenv@0.1.0, ini@1.3.3, once@1.3.1, nopt@3.0.1, config-chain@1.1.8, semver@4.3.3)
├── gaze@0.5.1 (globule@0.1.0)
├── sass-graph@1.3.0 (commander@2.8.1, glob@4.5.3, lodash@2.4.1)
├── request@2.55.0 (caseless@0.9.0, json-stringify-safe@5.0.0, aws-sign2@0.5.0, forever-agent@0.6.1, stringstream@0.0.4, oauth-sign@0.6.0, tunnel-agent@0.4.0, isstream@0.1.2, node-uuid@1.4.3, qs@2.4.1, combined-stream@0.0.7, form-data@0.2.0, mime-types@2.0.10, http-signature@0.10.1, bl@0.9.4, tough-cookie@0.13.0, hawk@2.3.1, har-validator@1.6.1)
└── pangyp@2.1.0 (which@1.0.9, osenv@0.1.0, rimraf@2.2.8, graceful-fs@3.0.6, nopt@3.0.1, glob@4.3.5, semver@4.2.2, fstream@1.0.4, minimatch@2.0.4, tar@1.0.3, npmlog@1.0.0, request@2.51.0)
````

**Comment by saper** [comment]
Regarding your compilation problems: which version of clang are you using? Do you have gcc as the system compiler? (what is your 'cc')? 

I have node-sass compiling fine with systems as old as FreeBSD 8.4 (I am using gcc 4.8 to get c++11 faciities), so no problem here as well...

**Comment by vladyslav2** [comment]
```
# clang --version
FreeBSD clang version 3.3 (tags/RELEASE_33/final 183502) 20130610
Target: x86_64-unknown-freebsd10.0
Thread model: posix
```
Compiling with clang finished with error:

```
root@computer# # env CC=clang CXX=clang++ LINK=clang++ SKIP_SASS_BINARY_DOWNLOAD_FOR_CI=1 V=1 npm -g install node-sass@beta

../node_modules/nan/nan.h:143:5: error: use of undeclared identifier 'abort'
    abort();
    ^
../node_modules/nan/nan.h:152:5: error: use of undeclared identifier 'abort'
    abort();
    ^
../src/binding.cpp:13:25: warning: unused variable 'ctx_w' [-Wunused-variable]
  sass_context_wrapper* ctx_w = static_cast<sass_context_wrapper*>(cookie);
```

```
# gcc49 --version
gcc49 (FreeBSD Ports Collection) 4.9.2 20141029 (prerelease)
Copyright (C) 2014 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

Compiling finished with error:

```
root@computer# env CC=gcc49 CXX=gcc49 LINK=gcc49 SKIP_SASS_BINARY_DOWNLOAD_FOR_CI=1 V=1 npm -g install node-sass@beta

../src/libsass/util.cpp: In function 'char* Sass::sass_strdup(const char*)':
../src/libsass/util.cpp:10:15: error: 'stderr' was not declared in this scope
       fprintf(stderr, "Out of memory.\n");    \
               ^
../src/libsass/util.cpp:19:7: note: in expansion of macro 'out_of_memory'
       out_of_memory();
       ^
../src/libsass/util.cpp:10:41: error: 'fprintf' was not declared in this scope
       fprintf(stderr, "Out of memory.\n");    \
                                         ^
../src/libsass/util.cpp:19:7: note: in expansion of macro 'out_of_memory'
       out_of_memory();
       ^
```

```
# gcc47 --version
gcc47 (FreeBSD Ports Collection) 4.7.4
Copyright (C) 2012 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

Does not showed any error on installation, here is last rows:

```
root@computer# env CC=gcc47 CXX=gcc47 LINK=gcc47 SKIP_SASS_BINARY_DOWNLOAD_FOR_CI=1 V=1 npm -g install node-sass@beta

 Release/obj.target/binding/src/sass_context_wrapper.o Release/obj.target/binding/src/sass_types/boolean.o Release/obj.target/binding/src/sass_types/color.o Release/obj.target/binding/src/sass_types/error.o Release/obj.target/binding/src/sass_types/factory.o Release/obj.target/binding/src/sass_types/list.o Release/obj.target/binding/src/sass_types/map.o Release/obj.target/binding/src/sass_types/null.o Release/obj.target/binding/src/sass_types/number.o Release/obj.target/binding/src/sass_types/string.o Release/obj.target/src/sass.a -Wl,--end-group 
  rm -rf "Release/binding.node" && cp -af "Release/obj.target/binding.node" "Release/binding.node"
gmake: Leaving directory '/usr/local/lib/node_modules/node-sass/build'
Installed in ` /usr/local/lib/node_modules/node-sass/vendor/freebsd-x64-11/binding.node `
node-sass@3.0.0-beta.7 /usr/local/lib/node_modules/node-sass
├── get-stdin@4.0.1
├── async-foreach@0.1.3
├── nan@1.8.2
├── mkdirp@0.5.0 (minimist@0.0.8)
├── chalk@1.0.0 (escape-string-regexp@1.0.3, ansi-styles@2.0.1, supports-color@1.3.1, strip-ansi@2.0.1, has-ansi@1.0.3)
├── meow@3.1.0 (object-assign@2.0.0, minimist@1.1.1, camelcase-keys@1.0.0, indent-string@1.2.1)
├── gaze@0.5.1 (globule@0.1.0)
├── glob@5.0.5 (path-is-absolute@1.0.0, inherits@2.0.1, once@1.3.1, minimatch@2.0.4, inflight@1.0.4)
├── npmconf@2.1.1 (uid-number@0.0.5, inherits@2.0.1, osenv@0.1.0, ini@1.3.3, semver@4.3.3, config-chain@1.1.8, nopt@3.0.1, once@1.3.1)
├── sass-graph@1.3.0 (commander@2.8.1, lodash@2.4.1, glob@4.5.3)
├── request@2.55.0 (caseless@0.9.0, json-stringify-safe@5.0.0, aws-sign2@0.5.0, forever-agent@0.6.1, stringstream@0.0.4, oauth-sign@0.6.0, tunnel-agent@0.4.0, isstream@0.1.2, node-uuid@1.4.3, qs@2.4.1, combined-stream@0.0.7, form-data@0.2.0, mime-types@2.0.10, http-signature@0.10.1, bl@0.9.4, tough-cookie@0.13.0, hawk@2.3.1, har-validator@1.6.1)
└── pangyp@2.1.0 (which@1.0.9, osenv@0.1.0, rimraf@2.2.8, graceful-fs@3.0.6, nopt@3.0.1, semver@4.2.2, fstream@1.0.4, minimatch@2.0.4, glob@4.3.5, tar@1.0.3, npmlog@1.0.0, request@2.51.0) 
```


But when i want to see installed modules:

```
root@computer:/home/vladka/webdevelop.pro/files/markup/exm# npm list
foundation-libsass-template@0.0.1 /home/vladka/webdevelop.pro/files/markup/exm
└── (empty)
```

I'm not seeing node-sass here ... 
is it okay ?

**Comment by vladyslav2** [comment]
Its too strange - gcc47 tell that node-sass was installed into ls /usr/local/lib/node_modules/, but:

```
# ls /usr/local/lib/node_modules/
.keepme                     bower                       foundation-libsass-template grunt-cli                   npm
```

```
# npm install grunt-cli
# npm install grunt-sass

# grunt
Loading "sass.js" tasks...ERROR
>> Error: `libsass` bindings not found. Try reinstalling `node-sass`?
>> Local Npm module "grunt-contrib-watch" not found. Is it installed?
Warning: Task "watch" not found. Use --force to continue.

Aborted due to warnings.
```

that strange

**Comment by saper** [comment]
Sorry, I am out of ideas.

**Comment by saper** [comment]
By the way in the gcc49 try using g++49 for the compiler and the linker.

**Comment by kkoopa** [comment]
libsass does not include cstdio in the compilation unit util.cpp, that causes errors with undefined `fprintf` and `stderr`. NAN did not include cstdlib in version 1.8.0, but it is fixed in 1.8.4. node-sass depends on nan: ^1.7.0, which covers 1.8.4, so that is already fixed, what remains is fixing libsass to include the cstdio header.

**Comment by vladyslav2** [comment]
@kkoopa so i just need to wait ?

**Comment by xzyfer** [comment]
I'll bump the nan dep to ^1.8.4 to prevent transient install issues.

**Comment by xzyfer** [comment]
I'm more concern as to why the install script needs to compile the binary.
The precompiled binary should do the job.

It doesn't even look like it tried to download when run as root.

**Comment by kkoopa** [comment]
@vladyslav2 Yes, it would need fixing upstream, but if you don't want to wait you could always apply the patch yourself.

**Comment by xzyfer** [comment]
The missing headers has been fixed in [3.2.1](https://github.com/sass/libsass/releases/tag/3.2.1)

**Comment by xzyfer** [comment]
Fixed in 3.0.0.
